### DISCUSSION TITLE
[Run issue]:How to Reduce NTuple Scorer Memory Requirements in TOPAS

### DISCUSSION THREAD

**Discussion by WKjie** [OpenTOPAS/OpenTOPAS#107] [Run-Problem]
### Contact Details

weikaijie644@gmail.com

### What was the issue?

Hi ,
I'm currently using the NTuple scorer in my TOPAS simulations, but I'm encountering memory limitations on my computer that cause the simulation to fail and my system to become unresponsive.
Could anyone suggest ways to reduce the memory requirements of the NTuple scorer? I've considered the following approaches but am unsure how to implement them in TOPAS:
1. Data Filtering: Is it possible to set conditions to filter out unnecessary events before recording?
2. Downsampling: Can certain events be ignored under specific conditions to decrease data volume?
3. Increase Step Size: Is it feasible to increase the step size to reduce simulation precision and thus decrease data volume?
4. Parameter Adjustment: How can I adjust parameters to reduce the range or precision of recorded data?
Any advice or examples on how to effectively implement these or other strategies in TOPAS would be greatly appreciated.
Thank you!

### TOPAS category

Scoring

### Version

4.0

### What Operative System (OS) are you running on?

Linux

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Attestation

- [X] I testify that I have fully read the bug reporting guideline beforehand.

**Comment by jschuemann** [comment]
Hi,
you can do all of these things. Changing the MaxStepSize may help a bit, but most times our simulation steps are physics limited rather than step size limited, so this may only help a little bit.
For most of the other options, you will need to edit the ntuple scoring function. Then you can set any rules for when you actually record an event either by adding limits or selecting or grouping events. Basically, you can define what you want to score there. You'll have to recompile TOPAS again and then should be set to try.
But are you running out of hard disk memory? Then maybe just consider running less events as well? How big do your files get? ntuples can get large quickly, but if you have many GB of data, maybe you can optimize your simulation.

**Reply by WKjie** [reply]
Thank you for your explanation, sir. Are you suggesting that I modify the `.cc` file for the Ntuple? My understanding of C++ is quite limited. Below is my setup for the Ntuple in Topas. Should I modify some of these parameters or alter other default parameters it contains to reduce memory usage?
d:Sc/pTuple/TimeCut 			= 1.0 ns
b:Sc/pTuple/IncludeChemicalTrack 	= "false"
s:Sc/pTuple/Quantity 			= "Tuple"
s:Sc/pTuple/Component 			= "Nucleus"
s:Sc/pTuple/OutputType 			= "ASCII"
b:Sc/pTuple/OutputToConsole 		= "True"
s:Sc/pTuple/IfOutputFileAlreadyExists 	= "overwrite"
b:Sc/pTuple/OutputAfterRun 		= "True"
b:Sc/pTuple/IncludeVolumeName		= "True"
b:Sc/pTuple/IncludeVolumeCopyNumber	= "True"
b:Sc/pTuple/IncludeEventID 		= "True"
b:Sc/pTuple/IncludeEnergyDeposited 	= "True"
b:Sc/pTuple/IncludePhysicalProcessName 	= "True"
b:Sc/pTuple/PropagateToChildren		= "True"
s:Sc/pTuple/OutputFile = "TupleFileName_Seed1"

**Reply by tmasilela** [reply]
Hi,

Yes. Depending on what information you're fine excluding, some of those parameters can be changed to "False". Alternatively you can use the "Binary" format which is less memory intensive than "ASCII". Other solutions would require modification of the source code.

All the best,
Thongchai
